HW2/P4/driver.py

```diff
@@ -27,6 +27,32 @@ def py_median_3x3(image, iterations=10, num_threads=1):
 
     return tmpA
 
+
+def py_median_3x3_multithread(image, iteration=10, num_threads=1):
+    tempA = image.copy()
+    tempB = np.empty_like(tempA)
+
+    for i in range(iteration):
+        # initiate the thread_list
+        thread_list = []
+        for j in range(num_threads):
+            # create new thread to do the filtering
+            new_thread = threading.Thread(target = filtering.median_3x3,
+                                          args = (tempA, tempB, j, num_threads))
+            new_thread.start()
+            thread_list.append(new_thread)
+
+        # join all the threads together
+        for thread in thread_list:
+            thread.join()
+        # swap direction of filtering
+        tempA, tempB = tempB, tempA
```

Comment:
I am not sure if you are doing what we were asked to do... the instructions say: "Use some combination of Semaphores, Condition Variables, Events, or other communication primitives from the Python threading module to ensure threads do not start an iteration of filtering before the data for that iteration is ready."
How are you ensuring that threads do not start an iteration before that data is ready? It does not seem like they are communicating, or I am not getting something?